File: zig/participant-config/subscriber.zig
//! zig/participant-config -- subscriber.
//!
//! Same two mutually exclusive modes as the publisher (see publisher.zig's
//! doc comment): programmatic (default) round-trips a config value through
//! set_default_participant_config/get_default_participant_config before
//! creating the participant via create_participant_ex; file mode
//! (--config <path>) loads a zzdds.toml-style file instead. Both then run
//! the same minimal reliable read loop (3 samples) as zig/hello_world.
//!
//! Required stdout markers: "Create topic:", "Create reader for topic:",
//! "Subscriber: received count=", "Subscriber: received all 3 samples in
//! order." Programmatic mode additionally prints "Config round-trip OK: ...".
//! Any failure path prints a line starting "FAIL:" and exits nonzero.

const std = @import("std");
const zzdds = @import("zzdds");
const DDS = @import("zzdds_generated").DDS;
const ZZDDS = @import("zzdds_ext_generated").zzdds;
const ping_gen = @import("config_ping_gen");

const EXPECTED_SAMPLES: i32 = 3;
const RECEIVE_TIMEOUT_NS: i64 = 30 * std.time.ns_per_s;
const POLL_PERIOD_NS: u64 = 20 * std.time.ns_per_ms;

const CONFIG_PARTICIPANT_NAME = "participant-config-example";
const CONFIG_FRAGMENT_SIZE: u16 = 9000;

fn monoNs(io: std.Io) i64 {
    return @intCast(std.Io.Clock.awake.now(io).nanoseconds);
}

fn sleepNs(io: std.Io, ns: u64) void {
    (std.Io.Clock.Duration{ .raw = .{ .nanoseconds = @intCast(ns) }, .clock = .awake }).sleep(io) catch {};
}

const State = struct {
    expected_next: i32 = 0,
    all_received: std.atomic.Value(bool) = std.atomic.Value(bool).init(false),
    alloc: std.mem.Allocator,
};

fn onDataAvailable(state: *State, dr: DDS.DataReader) void {
    var reader = ping_gen.ConfigPingDataReader.init(dr, state.alloc);
    while (true) {
        var value: ping_gen.ConfigPing = .{};
        var info: DDS.SampleInfo = .{};
        const got = reader.take_next_sample(&value, &info) catch {
            std.debug.print("FAIL: take_next_sample() CDR error\n", .{});
            std.process.exit(1);
        };
        if (!got) break;
        if (!info.valid_data) continue;

        if (value.count != state.expected_next) {
            std.debug.print("FAIL: expected count={d} but got count={d}\n", .{ state.expected_next, value.count });
            std.process.exit(1);
        }

        std.debug.print("Subscriber: received count={d}\n", .{value.count});
        state.expected_next += 1;

        if (state.expected_next == EXPECTED_SAMPLES) {
            state.all_received.store(true, .release);
        }
    }
}

const Options = struct {
    domain_id: u32 = 0,
    config_path: ?[]const u8 = null,
};

fn parseArgs(process_args: std.process.Args) Options {
    var opts = Options{};
    var it = std.process.Args.Iterator.init(process_args);
    _ = it.skip(); // program name
    while (it.next()) |arg| {
        if (std.mem.eql(u8, arg, "-d") or std.mem.eql(u8, arg, "--domain")) {
            const v = it.next() orelse continue;
            opts.domain_id = std.fmt.parseInt(u32, v, 10) catch 0;
        } else if (std.mem.eql(u8, arg, "--config")) {
            opts.config_path = it.next();
        }
    }
    return opts;
}

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    var gpa = std.heap.DebugAllocator(.{}){};
    defer _ = gpa.deinit();
    const alloc = gpa.allocator();

    const opts = parseArgs(init.minimal.args);

    if (opts.config_path) |path| {
        zzdds.process_config.configureFromFile(std.heap.c_allocator, path) catch |err| {
            std.debug.print("FAIL: failed to load config file '{s}': {}\n", .{ path, err });
            std.process.exit(1);
        };
    }

    var factory = zzdds.createFactory() catch {
        std.debug.print("FAIL: createFactory() failed\n", .{});
        std.process.exit(1);
    };
    defer factory.deinit();

    var dp: DDS.DomainParticipant = undefined;
    if (opts.config_path == null) {
        var cfg: ZZDDS.DomainParticipantConfig = .{};
        cfg.participant.name = CONFIG_PARTICIPANT_NAME;
        cfg.rtps.fragment_size = CONFIG_FRAGMENT_SIZE;

        const zfactory = factory.toZZDDSFactory();
        if (zfactory.set_default_participant_config(cfg) != DDS.RETCODE_OK) {
            std.debug.print("FAIL: set_default_participant_config() failed\n", .{});
            std.process.exit(1);
        }

        var readback: ZZDDS.DomainParticipantConfig = std.mem.zeroes(ZZDDS.DomainParticipantConfig);
        if (zfactory.get_default_participant_config(&readback) != DDS.RETCODE_OK) {
            std.debug.print("FAIL: get_default_participant_config() failed\n", .{});
            std.process.exit(1);
        }
        defer readback.deinit(std.heap.c_allocator);

        if (!std.mem.eql(u8, readback.participant.name, CONFIG_PARTICIPANT_NAME)) {
            std.debug.print("FAIL: participant.name round-trip mismatch: expected '{s}', got '{s}'\n", .{ CONFIG_PARTICIPANT_NAME, readback.participant.name });
            std.process.exit(1);
        }
        if (readback.rtps.fragment_size != CONFIG_FRAGMENT_SIZE) {
            std.debug.print("FAIL: rtps.fragment_size round-trip mismatch: expected {d}, got {d}\n", .{ CONFIG_FRAGMENT_SIZE, readback.rtps.fragment_size });
            std.process.exit(1);
        }
        std.debug.print("Config round-trip OK: participant.name='{s}' rtps.fragment_size={d}\n", .{ readback.participant.name, readback.rtps.fragment_size });

        dp = zfactory.create_participant_ex(opts.domain_id, DDS.DomainParticipantQos{}, null, 0, cfg);
    } else {
        const dpf = factory.toDDSFactory();
        dp = dpf.create_participant(opts.domain_id, .{}, null, 0);
    }
    if (dp.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_participant() failed on domain {d}\n", .{opts.domain_id});
        std.process.exit(1);
    }
    defer _ = factory.toDDSFactory().delete_participant(dp);

    var ts_alloc = alloc;
    if (!zzdds.registerTypeSupport(dp, "ConfigPing", .{
        .ctx = @ptrCast(&ts_alloc),
        .compute_key_hash = ping_gen.ConfigPing.computeKeyHashFromCdr,
    })) {
        std.debug.print("FAIL: registerTypeSupport() failed\n", .{});
        std.process.exit(1);
    }

    const topic = dp.create_topic("ConfigPing", "ConfigPing", .{}, null, 0);
    if (topic.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_topic() failed\n", .{});
        std.process.exit(1);
    }
    std.debug.print("Create topic: ConfigPing\n", .{});

    const subscriber = dp.create_subscriber(.{}, null, 0);
    if (subscriber.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_subscriber() failed\n", .{});
        std.process.exit(1);
    }

    var dr_qos = DDS.DataReaderQos{};
    dr_qos.reliability.kind = .RELIABLE_RELIABILITY_QOS;
    dr_qos.history.kind = .KEEP_ALL_HISTORY_QOS;

    var state = State{ .alloc = alloc };
    const dr_listener = DDS.dataReaderListener(&state, .{
        .on_data_available = onDataAvailable,
    });

    const topic_desc = dp.lookup_topicdescription("ConfigPing");
    const dr = subscriber.create_datareader(topic_desc, dr_qos, dr_listener, DDS.DATA_AVAILABLE_STATUS);
    if (dr.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_datareader() failed\n", .{});
        std.process.exit(1);
    }
    std.debug.print("Create reader for topic: ConfigPing\n", .{});

    std.debug.print("Subscriber: waiting for {d} samples...\n", .{EXPECTED_SAMPLES});
    const deadline = monoNs(io) + RECEIVE_TIMEOUT_NS;
    while (!state.all_received.load(.acquire)) {
        if (monoNs(io) > deadline) {
            std.debug.print("FAIL: only received {d}/{d} samples within 30s\n", .{ state.expected_next, EXPECTED_SAMPLES });
            std.process.exit(1);
        }
        sleepNs(io, POLL_PERIOD_NS);
    }

    _ = subscriber.delete_datareader(dr);

    std.debug.print("Subscriber: received all {d} samples in order.\n", .{EXPECTED_SAMPLES});
}

File: java/participant-config/Subscriber.java
// java/participant-config -- subscriber. Direct Java port of
// zig/participant-config/subscriber.zig / c/participant-config/src/subscriber.c
// / cpp/participant-config/src/subscriber.cpp; see
// docs/design/participant-config-reference-app.md at the repo root for the
// full spec. Same two mutually exclusive modes as the publisher (see
// Publisher.java's doc comment).
//
// Required stdout markers: "Create topic:", "Create reader for topic:",
// "Subscriber: received count=", "Subscriber: received all 3 samples in
// order." Programmatic mode additionally prints "Config round-trip OK: ...".
// Any failure path prints a line starting "FAIL:" and exits nonzero.

import io.zzdds.dcps.Dcps;
import io.zzdds.ext.Zzdds;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

public class Subscriber {
    static final int EXPECTED_SAMPLES = 3;
    static final int RECEIVE_TIMEOUT_MS = 30000;
    static final int POLL_PERIOD_MS = 20;

    static final String CONFIG_PARTICIPANT_NAME = "participant-config-example";
    static final short CONFIG_FRAGMENT_SIZE = 9000;

    static class Options {
        int domainId = 0;
        String configPath = null;
    }

    static Options parseArgs(String[] args) {
        Options opts = new Options();
        for (int i = 0; i < args.length; i++) {
            if ((args[i].equals("-d") || args[i].equals("--domain")) && i + 1 < args.length) {
                opts.domainId = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--config") && i + 1 < args.length) {
                opts.configPath = args[++i];
            }
        }
        return opts;
    }

    public static void main(String[] args) throws Exception {
        Options opts = parseArgs(args);

        if (opts.configPath != null) {
            int cfgRc = io.zzdds.runtime.ZzddsRuntime.configureFromFile(opts.configPath);
            if (cfgRc != Dcps.DDS.RETCODE_OK.value) {
                System.err.println("FAIL: failed to load config file '" + opts.configPath + "' (rc=" + cfgRc + ")");
                System.exit(1);
            }
        }

        Object baseFactory = io.zzdds.runtime.ZzddsRuntime.createFactory();
        if (baseFactory == null) {
            System.err.println("FAIL: createFactory() failed");
            System.exit(1);
        }

        Dcps.DDS.DomainParticipant dp;
        if (opts.configPath == null) {
            Zzdds.zzdds.DomainParticipantFactory factory =
                (Zzdds.zzdds.DomainParticipantFactory) io.zzdds.runtime.ZzddsRuntime.asZzddsFactory(baseFactory);
            if (factory == null) {
                System.err.println("FAIL: asZzddsFactory() failed");
                System.exit(1);
            }

            Zzdds.zzdds.DomainParticipantConfig cfg = new Zzdds.zzdds.DomainParticipantConfig();
            cfg.get_participant().set_name(CONFIG_PARTICIPANT_NAME);
            cfg.get_rtps().set_fragment_size(CONFIG_FRAGMENT_SIZE);

            if (factory.set_default_participant_config(cfg) != Dcps.DDS.RETCODE_OK.value) {
                System.err.println("FAIL: set_default_participant_config() failed");
                System.exit(1);
            }

            Zzdds.zzdds.DomainParticipantConfig readback = new Zzdds.zzdds.DomainParticipantConfig();
            if (factory.get_default_participant_config(readback) != Dcps.DDS.RETCODE_OK.value) {
                System.err.println("FAIL: get_default_participant_config() failed");
                System.exit(1);
            }

            if (!CONFIG_PARTICIPANT_NAME.equals(readback.get_participant().get_name())) {
                System.err.println("FAIL: participant.name round-trip mismatch: expected '" + CONFIG_PARTICIPANT_NAME
                        + "', got '" + readback.get_participant().get_name() + "'");
                System.exit(1);
            }
            if (readback.get_rtps().get_fragment_size() != CONFIG_FRAGMENT_SIZE) {
                System.err.println("FAIL: rtps.fragment_size round-trip mismatch: expected " + CONFIG_FRAGMENT_SIZE
                        + ", got " + readback.get_rtps().get_fragment_size());
                System.exit(1);
            }
            System.out.println("Config round-trip OK: participant.name='" + readback.get_participant().get_name()
                    + "' rtps.fragment_size=" + readback.get_rtps().get_fragment_size());

            dp = factory.create_participant_ex(opts.domainId, null, null, 0, cfg);
        } else {
            Dcps.DDS.DomainParticipantFactory factory = (Dcps.DDS.DomainParticipantFactory) baseFactory;
            dp = factory.create_participant(opts.domainId, null, null, 0);
        }
        if (dp == null) {
            System.err.println("FAIL: create_participant() failed on domain " + opts.domainId);
            System.exit(1);
        }

        if (ConfigPingTypeSupport.register(dp, "ConfigPing") != 0) {
            System.err.println("FAIL: register_type_support failed");
            System.exit(1);
        }

        Dcps.DDS.Topic topic = dp.create_topic("ConfigPing", "ConfigPing", null, null, 0);
        if (topic == null) {
            System.err.println("FAIL: create_topic() failed");
            System.exit(1);
        }
        System.out.println("Create topic: ConfigPing");

        Dcps.DDS.Subscriber sub = dp.create_subscriber(null, null, 0);
        if (sub == null) {
            System.err.println("FAIL: create_subscriber() failed");
            System.exit(1);
        }

        Dcps.DDS.DataReaderQos drQos = new Dcps.DDS.DataReaderQos();
        sub.get_default_datareader_qos(drQos);
        drQos.get_reliability().set_kind(Dcps.DDS.ReliabilityQosPolicyKind.RELIABLE_RELIABILITY_QOS);
        drQos.get_history().set_kind(Dcps.DDS.HistoryQosPolicyKind.KEEP_ALL_HISTORY_QOS);

        final AtomicInteger expectedNext = new AtomicInteger(0);
        final AtomicBoolean allReceived = new AtomicBoolean(false);
        final Object[] readerBox = new Object[1]; // set right after create_datareader

        Dcps.DDS.DataReaderListener listener = new Dcps.DDS.DataReaderListener() {
            public void on_requested_deadline_missed(Dcps.DDS.DataReader r, Dcps.DDS.RequestedDeadlineMissedStatus s) {}
            public void on_requested_incompatible_qos(Dcps.DDS.DataReader r, Dcps.DDS.RequestedIncompatibleQosStatus s) {}
            public void on_sample_rejected(Dcps.DDS.DataReader r, Dcps.DDS.SampleRejectedStatus s) {}
            public void on_liveliness_changed(Dcps.DDS.DataReader r, Dcps.DDS.LivelinessChangedStatus s) {}
            public void on_subscription_matched(Dcps.DDS.DataReader r, Dcps.DDS.SubscriptionMatchedStatus s) {}
            public void on_sample_lost(Dcps.DDS.DataReader r, Dcps.DDS.SampleLostStatus s) {}

            public void on_data_available(Dcps.DDS.DataReader r) {
                ConfigPingDataReader reader = (ConfigPingDataReader) readerBox[0];
                ConfigPingDataReader.Sample sample;
                while ((sample = reader.take()) != null) {
                    if (!sample.validData) continue;

                    int count = sample.data.get_count();
                    int expected = expectedNext.get();
                    if (count != expected) {
                        System.err.println("FAIL: expected count=" + expected + " but got count=" + count);
                        System.exit(1);
                    }

                    System.out.println("Subscriber: received count=" + count);
                    int next = expectedNext.incrementAndGet();

                    if (next == EXPECTED_SAMPLES) {
                        allReceived.set(true);
                    }
                }
            }
        };

        Dcps.DDS.DataReader rawReader = sub.create_datareader(topic, drQos, listener, Dcps.DDS.DATA_AVAILABLE_STATUS.value);
        if (rawReader == null) {
            System.err.println("FAIL: create_datareader() failed");
            System.exit(1);
        }
        System.out.println("Create reader for topic: ConfigPing");

        readerBox[0] = new ConfigPingDataReader(rawReader);

        System.out.println("Subscriber: waiting for " + EXPECTED_SAMPLES + " samples...");
        long deadline = System.currentTimeMillis() + RECEIVE_TIMEOUT_MS;
        while (!allReceived.get()) {
            if (System.currentTimeMillis() > deadline) {
                System.err.println("FAIL: only received " + expectedNext.get() + "/" + EXPECTED_SAMPLES + " samples within " + (RECEIVE_TIMEOUT_MS / 1000) + "s");
                System.exit(1);
            }
            Thread.sleep(POLL_PERIOD_MS);
        }

        sub.delete_datareader(rawReader);

        System.out.println("Subscriber: received all " + EXPECTED_SAMPLES + " samples in order.");
    }
}
